### PR TITLE
[R-package] add deprecation warnings on uses of '...' in predict() and reset_parameter()

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #4310 (based on https://github.com/microsoft/LightGBM/issues/4226#issuecomment-829473584).

This PR proposes adding some deprecation warnings on uses of `...` for parameters in the following functions:

* `predict.lgb.Booster()`
* `Booster$predict()`
* `Booster$reset_parameter()`

### Notes for reviewers

Removing `...` from `reset_parameter()` in release 4.0.0 would make that method's signature identical to the Python package's implementation.

https://github.com/microsoft/LightGBM/blob/4db10d86dcccc856cc42e9f66aef424590997f14/python-package/lightgbm/basic.py#L2870

However, removing `...` from the signature of `Booster$predict()` and `predict.lgb.Booster()` would be different from the behavior in the Python package, which allows additional parameters passed through `**kwargs`.

https://github.com/microsoft/LightGBM/blob/c65a2e3363cdff5ddb31689c6d555ec345932d98/python-package/lightgbm/basic.py#L3399-L3401

I think not allowing `...`, and instead encouraging the use of `params`, for additional prediction parameters is a better interface for the R package. It makes that interface consistent with all the other parts of the R package which use `params` (like Dataset construction and training), and leaves open the possibility of an improved user experience with "other" prediction parameters by implementing something like a prediction-parameters-specific version of the feature described in https://github.com/microsoft/LightGBM/issues/4195#issuecomment-822072038.

But @StrikerRUS , if you feel strongly that consistency with the Python package is more important than enabling that eventual "generate a `params` list from a function" feature for the specific case of `predict()` methods, I'd be ok with leaving the `...` for them.

### CODE REVIEW

**Comment by jameslamb** [comment]
I'll look into these CI failures on Mac and Linux builds tomorrow:

```text
predict.lgb.Booster: no visible global function definition for
  ‘modifyList’
Undefined global functions or variables:
  modifyList
Consider adding
  importFrom("utils", "modifyList")
to your NAMESPACE file.
```

Maybe `modifyList()` now needs to be explicitly imported and used with `::`. Very weird, given that it's used in several other places on `master` without causing any issues.

**Comment by jameslamb** [comment]
I just pushed https://github.com/microsoft/LightGBM/pull/4548/commits/799bd013d112a114752a78cfa559638c1b188ba3 to address https://github.com/microsoft/LightGBM/pull/4548#issuecomment-903455208

I'm still not sure why this NOTE is only showing up now, but that commit should fix it.

**Comment by StrikerRUS** [comment]
TBH, I'm not sure what consistency is better: consistency with the Python-package or consistency across the R-package. Probably the latter, but will defer to other maintainers' opinion.

File: R-package/R/lgb.Booster.R
Comment (StrikerRUS) [inline R-package/R/lgb.Booster.R:492]: > See ?predict.lgb.Booster for ...

Sorry, I'm not familiar with this, but it looks like a copy-paste typo from `predict.lgb.Booster`.

File: R-package/R/lgb.Booster.R
Comment (StrikerRUS) [inline R-package/R/lgb.Booster.R:483]: I suppose we should update all our tests (and examples, but if I'm not mistaken, we don't have any yet) and use `params` in them to silence new warning.

File: R-package/R/lgb.Booster.R
Comment (jameslamb) [reply R-package/R/lgb.Booster.R:492]: ah, good question! Nope, this one was not a mistake.

Since the `R6` class `Booster` is not currently exported https://github.com/microsoft/LightGBM/blob/11d7608f2d755899a88bfa1c0fbc779b9e9b3505/R-package/NAMESPACE, users of the R package cannot view any documentation for it with something like `?Booster`.

So I think that in the current state of the package, `?predict.lgb.Booster` is the best place to direct them.

File: R-package/R/lgb.Booster.R
Comment (jameslamb) [reply R-package/R/lgb.Booster.R:483]: Have you seen any such warnings or places where prediction parameters are passed through in our examples or tests? I checked through the tests and didn't find any uses of extra parameters that didn't match the other keyword arguments already.

To be clear, this would only apply for passing the parameters listed at https://lightgbm.readthedocs.io/en/latest/Parameters.html#predict-parameters, such as `pred_early_stop`.

I searched for those parameters with `git grep` and didn't find any uses in the R package. For example, with 

```shell
git grep pred_early_stop R-package/
```

**Comment by jameslamb** [comment]
> TBH, I'm not sure what consistency is better: consistency with the Python-package or consistency across the R-package. Probably the latter, but will defer to other maintainers' opinion.

Ok thank you! I think consistency within the R package is probably more important, and consistency between R and Python is valuable but not AS important.

So I think we should move forward with introducing this `params` pattern for extra prediction parameters in the R package.

File: R-package/R/lgb.Booster.R
Comment (StrikerRUS) [reply R-package/R/lgb.Booster.R:492]: Ah, OK, thanks a lot for the explnation!

File: R-package/R/lgb.Booster.R
Comment (StrikerRUS) [reply R-package/R/lgb.Booster.R:483]: Sorry, didn't know that you had already done it! Then everything is OK with this PR I think. But it's a good sign to add a test for those parameters in R-package 😉

**Review by StrikerRUS — APPROVED** [review]
LGTM, thanks!

File: R-package/R/lgb.Booster.R
Comment (jameslamb) [reply R-package/R/lgb.Booster.R:483]: No problem, I should have mentioned it. And definitely agree that it's a sign of a missing test.

**Pull request merged**
